### PR TITLE
new: Add colored log levels

### LINKED ISSUE #905: Add colors to log levels

### PR DESCRIPTION
**Pull request by camnwalter**
Closes #905 
Also see https://github.com/hermit-os/loader/pull/336

### CODE REVIEW

File: src/logging.rs [hunk 1]
```diff
@@ -14,8 +14,18 @@ impl log::Log for KernelLogger {
 
 	fn log(&self, record: &Record<'_>) {
 		if self.enabled(record.metadata()) {
+			let color_code: &str = match record.level() {
+				log::Level::Error => "\x1b[31m", // red
+				log::Level::Debug => "\x1b[32m", // green
+				log::Level::Trace => "\x1b[35m", // magenta
+				log::Level::Warn => "\x1b[33m",  // yellow
+				log::Level::Info => "\x1b[34m",  // blue
```
Comment (mkroening) [inline src/logging.rs:22]: Please import `log::Level` in cases like this.

**Review by mkroening — CHANGES_REQUESTED** [review]
Thanks a lot for the PR! :)

This is great, but I have a few visual and technical requests:

Visual requests:

1. I think we should only color the level (e.g., `WARN` in `[WARN]`) and not the whole line.
2. For color choice I would prefer if we would follow what tracing-subscriber is doing: [tracing-subscriber/src/fmt/format/mod.rs#L1504-L1510](https://github.com/tokio-rs/tracing/blob/tracing-subscriber-0.3.18/tracing-subscriber/src/fmt/format/mod.rs#L1504-L1510)
3. We should respect the [`NO_COLOR`](https://no-color.org/) environment variable at build time to disable any colors.

Technical requests:

I think it would be good, if we introduced a newtype `ColorLevel(Level)` that implements a colored 'Display' via [`anstyle`](/latest/anstyle/struct.Style.html).

**Comment by camnwalter** [comment]
> I think it would be good, if we introduced a newtype `ColorLevel(Level)` that implements a colored 'Display' via [`anstyle`](/latest/anstyle/struct.Style.html).

anstyle comes with the stdlib, so do you want me to re-create what they do?

**Comment by mkroening** [comment]
> anstyle comes with the stdlib, so do you want me to re-create what they do?

If you disable the default `std` feature, it should work:

```toml
anstyle = { version = "1", default-features = false }
```

**Review by mkroening — CHANGES_REQUESTED** [review]
> anstyle comes with the stdlib, so do you want me to re-create what they do?

If you disable the default `std` feature, it should work:

```toml
anstyle = { version = "1", default-features = false }
```

File: src/logging.rs [hunk 2]
```diff
@@ -17,13 +20,37 @@ impl log::Log for KernelLogger {
...
+
+impl fmt::Display for ColorLevel {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		let level: Level = self.0;
+		let color: AnsiColor = match level {
+			Level::Error => AnsiColor::Red,
+			Level::Debug => AnsiColor::Yellow,
+			Level::Trace => AnsiColor::Magenta,
+			Level::Info => AnsiColor::Green,
+			Level::Warn => AnsiColor::Blue,
```
Comment (mkroening) [inline src/logging.rs:40]:
```suggestion
			Level::Trace => AnsiColor::Magenta,
			Level::Debug => AnsiColor::Blue,
			Level::Info => AnsiColor::Green,
			Level::Warn => AnsiColor::Yellow,
			Level::Error => AnsiColor::Red,
```

File: src/logging.rs [hunk 3]
```diff
@@ -17,13 +20,37 @@ impl log::Log for KernelLogger {
...
+		};
+
+		let no_color: Option<&'static str> = option_env!("NO_COLOR");
+
+		if no_color.is_none() || no_color.unwrap().is_empty() {
```
Comment (mkroening) [inline src/logging.rs:45]: Let's extract this into a function:
```rust
fn color() -> bool {
    option_env!("NO_COLOR").is_some_and(|val| !val.is_empty())
}
```

**Review by mkroening — CHANGES_REQUESTED** [review]
Thanks, this is looking great. Could you remove the explicit types from the `let` bindings unless necessary? :)

File: src/logging.rs [hunk 4]
```diff
@@ -17,13 +20,40 @@ impl log::Log for KernelLogger {
...
+
+		if no_color() {
+			write!(f, "{level}")
+		} else {
+			let color = match level {
+				Level::Trace => AnsiColor::Magenta,
+				Level::Debug => AnsiColor::Yellow,
+				Level::Info => AnsiColor::Green,
+				Level::Warn => AnsiColor::Blue,
+				Level::Error => AnsiColor::Red,
```
Comment (mkroening) [inline src/logging.rs:44]:
```suggestion
				Level::Trace => AnsiColor::Magenta,
				Level::Debug => AnsiColor::Blue,
				Level::Info => AnsiColor::Green,
				Level::Warn => AnsiColor::Yellow,
				Level::Error => AnsiColor::Red,
```

**Review by mkroening — CHANGES_REQUESTED** [review]

File: src/logging.rs (hunk 4 above)
Comment (camnwalter) [reply src/logging.rs:44]: Oops I don't know what happened there!

**Review by mkroening — APPROVED** [review]
Thanks! :)

**Comment by camnwalter** [comment]
I'll do similar changes in the other pull request in a little bit

**Pull request merged**